File: examples/node/quickstart.js
/**
 * LiteVec Node.js Quickstart Example
 *
 * Demonstrates: create database → insert vectors → search → get results
 *
 * Requirements:
 *   npm install litevec
 */

const { Database } = require("litevec");

async function main() {
  // Open an in-memory database
  const db = Database.openMemory();

  // Create a collection for 128-dimensional vectors
  const collection = db.createCollection("documents", 128);

  // Sample vectors (in practice, these come from an embedding model)
  const documents = [
    {
      id: "doc1",
      vector: Array.from({ length: 128 }, (_, i) => Math.sin(i * 0.1)),
      metadata: JSON.stringify({ title: "Introduction to AI", category: "ai" }),
    },
    {
      id: "doc2",
      vector: Array.from({ length: 128 }, (_, i) => Math.cos(i * 0.1)),
      metadata: JSON.stringify({
        title: "Machine Learning Basics",
        category: "ml",
      }),
    },
    {
      id: "doc3",
      vector: Array.from({ length: 128 }, (_, i) =>
        Math.sin(i * 0.1 + Math.PI / 4)
      ),
      metadata: JSON.stringify({
        title: "Neural Networks",
        category: "deep-learning",
      }),
    },
  ];

  // Insert vectors
  console.log("Inserting vectors...");
  for (const doc of documents) {
    collection.insert(doc.id, doc.vector, doc.metadata);
  }
  console.log(`Inserted ${collection.len()} vectors\n`);

  // Search for similar vectors
  const query = Array.from({ length: 128 }, (_, i) => Math.sin(i * 0.1 + 0.1));
  console.log("Searching for similar vectors...");
  const results = collection.search(query, 3);

  console.log("Results:");
  for (const result of results) {
    const meta = JSON.parse(result.metadata || "{}");
    console.log(`  ${result.id}: distance=${result.distance.toFixed(4)}`);
    console.log(`    Title: ${meta.title || "N/A"}`);
  }

  // Get a specific vector
  console.log("\nGetting doc1...");
  const record = collection.get("doc1");
  if (record) {
    console.log(`  ID: ${record.id}`);
    console.log(`  Dimension: ${record.vector.length}`);
    console.log(`  Metadata: ${record.metadata}`);
  }

  // Delete a vector
  console.log("\nDeleting doc2...");
  const deleted = collection.delete("doc2");
  console.log(`  Deleted: ${deleted}`);
  console.log(`  Collection size: ${collection.len()}`);

  console.log("\n✅ Done!");
}

main().catch(console.error);

File: examples/rust/quickstart.rs
//! LiteVec Quickstart Example
//!
//! Run with: cargo run --example quickstart

use litevec::{Database, Filter};
use serde_json::json;

fn main() -> litevec::Result<()> {
    // Open an in-memory database (or use Database::open("my.lv") for persistence)
    let db = Database::open_memory()?;

    // Create a collection with 3-dimensional vectors
    let col = db.create_collection("documents", 3)?;

    // Insert vectors with metadata
    col.insert(
        "doc1",
        &[1.0, 0.0, 0.0],
        json!({"title": "Introduction to AI", "category": "tech"}),
    )?;
    col.insert(
        "doc2",
        &[0.0, 1.0, 0.0],
        json!({"title": "Machine Learning 101", "category": "tech"}),
    )?;
    col.insert(
        "doc3",
        &[0.0, 0.0, 1.0],
        json!({"title": "Cooking with Rust", "category": "food"}),
    )?;
    col.insert(
        "doc4",
        &[0.7, 0.7, 0.0],
        json!({"title": "Neural Networks Deep Dive", "category": "tech"}),
    )?;

    println!("Inserted {} vectors", col.len());

    // Search for the 3 nearest neighbors
    println!("\n--- Unfiltered Search ---");
    let results = col.search(&[0.9, 0.1, 0.0], 3).execute()?;
    for r in &results {
        println!(
            "  {} (distance: {:.4}) — {}",
            r.id, r.distance, r.metadata["title"]
        );
    }

    // Search with metadata filter
    println!("\n--- Filtered Search (category=tech) ---");
    let filtered = col
        .search(&[0.9, 0.1, 0.0], 3)
        .filter(Filter::Eq("category".into(), json!("tech")))
        .execute()?;
    for r in &filtered {
        println!(
            "  {} (distance: {:.4}) — {}",
            r.id, r.distance, r.metadata["title"]
        );
    }

    // Get a specific vector
    if let Some(record) = col.get("doc1")? {
        println!("\n--- Get doc1 ---");
        println!("  Vector: {:?}", record.vector);
        println!("  Metadata: {}", record.metadata);
    }

    // Delete a vector
    col.delete("doc3")?;
    println!("\nAfter delete: {} vectors remaining", col.len());

    // List collections
    println!("\nCollections: {:?}", db.list_collections());

    println!("\n✅ LiteVec quickstart complete!");
    Ok(())
}
